Alerting: Suggest previously entered custom labels

VikaCep opened the pull request:
**What is this feature?**

Instead of letting user write labels using plain input texts, we now display dropdowns that are filled with previously entered custom labels in other alert rules.


![2022-10-27 16 26 41](https://user-images.githubusercontent.com/6271380/198380837-c4cc870d-14d5-422f-aa22-9dc19ba804d2.gif)

Additionally, we show a tooltip indicating that the labels displayed in the dropdowns don't correspond to the entire set, but just the ones that the user previously entered in other alert rules. 

![image](https://user-images.githubusercontent.com/6271380/198381757-738659f6-e134-49dd-9385-efec37276ea8.png)

It's important to note that if the user wants to enter a new label that's not displayed in the list, then they can just type it and it will be added as usual:

![2022-10-27 16 34 16](https://user-images.githubusercontent.com/6271380/198382209-b717d0a9-5174-48a0-bd30-90851f26b7db.gif)



Fixes https://github.com/grafana/grafana/issues/45455

VikaCep committed:
[Alerting] - replace label inputs with dropdowns (#57019)

* Add AlertLabelDropdown component

It will be used to pick from or create new labels

* Adapt LabelsField component to use AlertLabelDropdown instead of inputs

* Add tests for LabelsField component

Plus a few other tests were adapted to work with the label dropdowns

* Use ref in component

* Fix showing placeholders in the label dropdowns

* Minor syntax change

VikaCep committed:
Display custom labels

When a label key is selected, its corresponding values are shown in the dropdown

VikaCep committed:
Add tooltip explaining where labels in the dropdowns come from

VikaCep committed:
Fix import of Stack component

VikaCep committed:
Avoid duplicated values

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -1,17 +1,65 @@
+
+  const rulerRuleRequests = useUnifiedAlertingSelector((state) => state.rulerRules);
+
+  const rulerRequest = rulerRuleRequests[GRAFANA_RULES_SOURCE_NAME];
```
What happens for Mimir/Cortex/Loki data sources? If we display for them labels from Grafana then users might be confused

soniaAguilarPeiron replied:
I think dataSourceName should be passed as a prop in LabelsField

VikaCep replied:
Do you think I should fetch rules for all the different datasources?

VikaCep replied:
I made some changes to only fetch the rules and display the labels for the current `dataSourceName` which is taken from the form context.

![2022-11-01 17 26 14](https://user-images.githubusercontent.com/6271380/199334352-943430b1-6f2d-4b17-970f-843a34ef678f.gif)

soniaAguilarPeiron replied:
If we want to have a re-usable `LabelField` we should not depend on the form that this `LabelField` is used. In contact points view, for example, we are using a different form.
So, I think that this `LabelField` component should receive `dataSource` as a prop, and the parent component should be the responsible for passing the correct value.

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -25,25 +73,49 @@ const LabelsField: FC<Props> = ({ className }) => {
+
+  const [selectedKey, setSelectedKey] = useState('');
 
   const keys = useMemo(() => {
```
If wonder if it's a good idea to display suggestions on the Test Contact Point modal. This is highly connected to what Alertmanager you use (it might be tricky to get the correct suggestions)

VikaCep replied:
Good point. I changed the `LabelsField` component to receive a `suggest` prop that controls whether to display the dropdowns or plain text inputs. 

`suggest=true`
![image](https://user-images.githubusercontent.com/6271380/199594959-cf4d12c6-4551-490e-8eb3-a295c430e4ca.png)

`suggest=false`
![image](https://user-images.githubusercontent.com/6271380/199595080-93c35d05-82f4-4575-a1dd-f1558b005996.png)

soniaAguilarPeiron replied:
Nice! Could we re-use code between `LabelsWithoutSuggestions` and `LabelsWithSuggestions`? I see a lot of common code in both components.

VikaCep replied:
Totally 👍

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -1,17 +1,65 @@
+
+  labels.forEach((label: Record<string, string>) => {
+    Object.entries(label).forEach(([key, value]) => {
+      labelsByKey[key] = uniq([...(labelsByKey[key] || []), value]);
```
nit: How about using `Set` to keep labels unique?

VikaCep replied:
Good idea 👍

VikaCep committed:
Improvements based on review

VikaCep committed:
Display labels for currently selected datasource only

VikaCep committed:
Refactor AlertsField to allow to choose whether to suggest labels or not

VikaCep committed:
Suggest labels for NotificationStep and tests

VikaCep committed:
Don't suggest labels in TestContactPointModal

soniaAguilarPeiron commented:
I'm thinking about having lots of alert rules, and as a consequence, loading ruler rules could take a long time... I think we should show some 'Loading...' spinner in the labels fields?

soniaAguilarPeiron commented:
Another point to take in account, is that if we want to have a re-usable `LabelField` we should not depend on the form that this `LabelField` is used. In contact points view, for example,  we are using a different form.
So, I think that this `LabelField` component should receive `dataSource` as a prop, and the parent component should be the responsible for passing the correct value.

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -23,94 +71,209 @@ const LabelsField: FC<Props> = ({ className }) => {
-      .map((field) => ({ label: field, value: field }));
+  const dataSourceName = watch('dataSourceName');
+
+  const labelsByKey = useGetCustomLabels(dataSourceName || RuleFormType.grafana);
```
can we relay on this condition? are we sure to have dataSourceName in the form in case of not having a grafana managed alert rule?

VikaCep replied:
Yes, I tried both in grafana managed alert rules and external alert managers. The datasource name is present in both cases. But I will do as you suggested and send it as prop from the parent component.

VikaCep committed:
[LabelsField] - refactor: get dataSourceName as a parameter

VikaCep committed:
[LabelsField] - extract common code into reusable components

VikaCep committed:
Display loading spinner while fetching rules

VikaCep commented:
> I'm thinking about having lots of alert rules, and as a consequence, loading ruler rules could take a long time... I think we should show some 'Loading...' spinner in the labels fields?

Good idea! I added a spinner while the rules are being fetched for the `LabelsWithSuggestions` component.
![image](https://user-images.githubusercontent.com/6271380/199776310-b11e9921-7cad-4f10-9155-80073600875c.png)

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -23,94 +113,173 @@ const LabelsField: FC<Props> = ({ className }) => {
+  );
+};
+
+const LabelsField: FC<Props> = ({ className, suggest, dataSourceName }) => {
```
I wonder if we should just have two separate components `LabelsWithSuggestions` and `LabelsWithoutSuggestions` and remove `LabelsField` wdyt? The one with suggestions does quite a lot of things under the hood compared to the one without suggestions so maybe we could just keep them separate

VikaCep replied:
I thought about this and I agreed with you, especially since I didn't like having both the `suggest` and `dataSourceName` props as they were too tied to each other. 
As I was about to implement it I thought of something else though: have the `dataSourceName` prop be optional in the `LabelsField` component and depending on its value/lack of, show either the dropdowns with suggestions or the text inputs if absent. This allowed me to remove the `suggest` prop. This way, we have the `LabelsField` component  in control of what to show based on whether it receives a `dataSourceName` or not. I kind of like having this logic encapsulated there.
What do you think? Let me know if you disagree as I have no problem changing it to two separate components though.

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -23,94 +113,173 @@ const LabelsField: FC<Props> = ({ className }) => {
-    [...new Set(fields.map((field) => field[type]))]
-      .filter((field) => Boolean(field))
-      .map((field) => ({ label: field, value: field }));
+  const { loading, labelsByKey } = useGetCustomLabels(dataSourceName || RuleFormType.grafana);
```
It seems to me we shouldn't fall back to `RuleTypeForm.grafana`. 
We should always require passing `dataSourceName` instead. `grafana` fallback might give us incorrect suggestions (creating/editing Mimir/Loki rules `grafana` suggestions are useless)

VikaCep replied:
You're right, I removed this fallback value. 
I left this prop as optional in the `LabelsField` component as based on this it either shows the dropdowns or the text inputs. The `LabelsWithSuggestions` component does make it mandatory now though. See https://github.com/grafana/grafana/pull/57783#discussion_r1015865847
Let me know what you think though :)

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -1,18 +1,108 @@
+    });
+  });
+
+  return { loading: false, labelsByKey };
```
Actually I think we could return directly in here loading from the request :

```
  const rulerRequest = rulerRuleRequests[dataSourceName];

  const result = rulerRequest.result || {};

  //store all labels in a flat array and remove empty values
  const labels = compact(
    flattenDeep(
      Object.keys(result).map((ruleGroupKey) =>
        result[ruleGroupKey].map((ruleItem: RulerRuleGroupDTO) => ruleItem.rules.map((item) => item.labels))
      )
    )
  );

  const labelsByKey: Record<string, string[]> = {};

  labels.forEach((label: Record<string, string>) => {
    Object.entries(label).forEach(([key, value]) => {
      labelsByKey[key] = [...new Set([...(labelsByKey[key] || []), value])];
    });
  });

  return { loading: rulerRequest.loading, labelsByKey: labelsByKey };
```

VikaCep replied:
Good idea!!

VikaCep committed:
LabelsField - refactor

Removing the suggest prop and the default dataSource 'grafana'. Instead, the component now relies on the dataSourceName param. If it's set it means we want to show suggestions so we fetch the labels, otherwise, if not set, we show the plain input texts without suggestions.

VikaCep committed:
Add test for LabelsField without suggestions

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rule-editor/NotificationsStep.tsx:
```diff
@@ -12,6 +15,10 @@ export const NotificationsStep = () => {
 
+  const { watch } = useFormContext<RuleFormValues>();
+
+  const dataSourceName = watch('dataSourceName');
```
With current changes , for grafana-managed alert rules you don't see the custom labels drop-down

<img width="1420" alt="Screenshot 2022-11-09 at 10 10 11" src="https://user-images.githubusercontent.com/33540275/200791310-b370f353-3db1-45af-a354-4567f94cca29.png">

this is because in case of grafana-managed alert rules this watch returns null, so, I think the correct dataSourceName should be:
`const dataSourceName = watch('dataSourceName') ?? GRAFANA_RULES_SOURCE_NAME`

VikaCep replied:
> With current changes , for grafana-managed alert rules you don't see the custom labels drop-down
> 
> <img alt="Screenshot 2022-11-09 at 10 10 11" width="1420" src="https://user-images.githubusercontent.com/33540275/200791310-b370f353-3db1-45af-a354-4567f94cca29.png">
> 
> this is because in case of grafana-managed alert rules this watch returns null, so, I think the correct dataSourceName should be: `const dataSourceName = watch('dataSourceName') ?? GRAFANA_RULES_SOURCE_NAME`

Great catch Sonia! 💯 You're right, for grafana managed alert rules we don't have the `dataSourceName` value. I agree it makes sense to set it ourselves if that's the case. Fix done 👍

VikaCep committed:
Show custom labels for grafana managed alerts

When the dataSourceName in the NotificationsStep component has a null value, we can assume it's because we're dealing with grafana managed alerts. In that case we set the correct value.

VikaCep committed:
Remove unneeded import after rebase

VikaCep committed:
Fix tests after latest changes

Since we removed the combobox from the TestContactPoints modal, tests had to be adjusted

VikaCep commented:
Hey @soniaAguilarPeiron and @konrad147, this PR is ready as I believe I addressed all your comments. Let me know otherwise 🙏

soniaAguilarPeiron reviewed (approved):
LGTM! nice!!🎉

VikaCep commented on public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx:
```diff
@@ -1,92 +1,276 @@
+          <Tooltip
+            content={
+              <div>
+                Labels shown as suggestions in the dropdowns come from previously entered labels in other alert rules.
+                They do not represent the full set of labels for the alert.
```
@brendamuir do you agree with this text? Or do you have any suggestions to change it? Basically this means we're not showing autogenerated labels but only the ones that were manually entered previously.

brendamuir replied:
The dropdown only displays labels that you have previously used for alerts. 
Select a label from the dropdown or type in a new one.

VikaCep committed:
Update texts

VikaCep committed:
initialize all new added inputs with empty data
